pyicloud/services/drive.py

```diff
@@ -53,6 +57,143 @@ def get_file(self, file_id, **kwargs):
         url = response.json()["data_token"]["url"]
         return self.session.get(url, params=self.params, **kwargs)
 
+    def _get_upload_contentws_url(self, file_object):
+        """Get the contentWS endpoint URL to add a new file."""
+        content_type = mimetypes.guess_type(file_object.name)[0]
+        if content_type is None:
+            content_type = ""
+
+        # Get filesize from file object
+        orig_pos = file_object.tell()
+        file_object.seek(0, os.SEEK_END)
+        file_size = file_object.tell()
+        file_object.seek(orig_pos, os.SEEK_SET)
+
+        file_params = self.params
+        file_params.update(self._get_token_from_cookie())
+
+        request = self.session.post(
+            self._document_root + "/ws/com.apple.CloudDocs/upload/web",
+            params=file_params,
+            headers={'Content-Type': 'text/plain'},
+            data=json.dumps(
+                {
+                    "filename": "%s" % file_object.name,
+                    "type": "FILE",
+                    "content_type": content_type,
+                    "size": "%i" % file_size,
+                }
+            ),
+        )
+        if not request.ok:
+            return None
+        return (request.json()[0]["document_id"], request.json()[0]["url"])
+
+    def _update_contentws(self, folder_id, sf_info, document_id, file_object):
+        request = self.session.post(
+            self._document_root + "/ws/com.apple.CloudDocs/update/documents",
+            params=self.params,
+            headers={'Content-Type': 'text/plain'},
+            data=json.dumps(
+                {
+                    "data": {
+                        "signature": "%s" % sf_info["fileChecksum"],
+                        "wrapping_key": "%s" % sf_info["wrappingKey"],
+                        "reference_signature": "%s" % sf_info["referenceChecksum"],
+                        "receipt": "%s" % sf_info["receipt"],
+                        "size": "%s" % sf_info["size"],
+                    },
+                    "command": "add_file",
+                    "create_short_guid": True,
+                    "document_id": "%s" % document_id,
+                    "path": {
+                        "starting_document_id": "%s" % folder_id,
+                        "path": "%s" % file_object.name,
```

Comment:
`"%s" % ` should not be nececassy

Reply:
Okay. Happy to change those. It was just the pattern I had seen in other places, so I thought I'd keep it.